server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java

```diff
@@ -35,12 +35,12 @@ private BlobPath(List<String> paths) {
     }
 
     public BlobPath add(String path) {
-        return new BlobPath(CollectionUtils.appendToCopy(this.paths, path));
+        return new BlobPath(CollectionUtils.appendToCopy(this.paths, path.endsWith("/") ? path.substring(0, path.length() - 1) : path));
     }
 
     public String buildAsString() {
         String p = String.join(SEPARATOR, paths);
-        if (p.isEmpty() || p.endsWith(SEPARATOR)) {
+        if (p.isEmpty()) {
             return p;
```

Comment:
I guess we can just remove an empty string here, because `p` would be empty only if `paths` is empty.

Reply:
Or we could just do

```java
if (paths.isEmpty()) {
    return "";
}
return paths.stream().collect(Collectors.joining(SEPARATOR, "", SEPARATOR));
```